### PR TITLE
Fix zeros_like for sparse tensors with batch dimensions. Add opinfo-based tests to like-functions.

### LINKED ISSUE #101078: `torch.zeros_like` produces wrong outputs with batched BSR/CSR inputs.
### 🐛 Describe the bug

BSR:
```python
In [1]: import torch

In [2]: x = torch.rand(2, 128, 128).to_sparse_bsr(16)

In [3]: y = torch.zeros_like(x)

In [4]: y.values().shape
Out[4]: torch.Size([0, 128, 16, 16])  # Ooops! Looks like 0 is in the wrong place.

```
CSR:
```python
In [1]: import torch

In [2]: x = torch.rand(2, 128, 128).to_sparse_csr()
<ipython-input-2-1147cb0239d1>:1: UserWarning: Sparse CSR tensor support is in beta state. If you miss a functionality in the sparse tensor support, please submit a feature request to https://github.com/pytorch/pytorch/issues. (Triggered internally at /home/nik/git/Quansight/pytorch/aten/src/ATen/SparseCsrTensorImpl.cpp:53.)
  x = torch.rand(2, 128, 128).to_sparse_csr()

In [3]: y = torch.zeros_like(x)

In [4]: y.values().shape  # No batch dim!
Out[4]: torch.Size([0, 128])

In [5]: y.crow_indices().shape  # No batch dim!
Out[5]: torch.Size([3])
```

### Versions

Current master.

cc @alexsamardzic @pearu @cpuhrsch @amjames @bhosmer

### PR DESCRIPTION
**Pull request by pearu**
Fixes #101078

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #100394
* __->__ #101215

### CODE REVIEW

File: torch/testing/_internal/opinfo/core.py [hunk 1]
```diff
@@ -1228,6 +1238,9 @@ def non_empty_sampler(op, generator):
             sample_inputs_mth(device, dtype, requires_grad=requires_grad, **kwargs),
         )
 
+    def _sample_inputs_unspecified(self, *args, **kwargs):
+        raise unittest.SkipTest("NO SAMPLE FUNCTION SPECIFIED!")
```
Comment (cpuhrsch) [inline torch/testing/_internal/opinfo/core.py:1242]: So we'll see a lot of skips from now on?

**Review by cpuhrsch — CHANGES_REQUESTED** [review]
Looks good to me, but I'm worried we're adding a ton of skips to our opinfos.

File: torch/testing/_internal/opinfo/core.py (hunk 1 above)
Comment (pearu) [reply torch/testing/_internal/opinfo/core.py:1242]: No, the number of such skips is relatively small (<100 out of tens of thousands sparse specific test tasks) and all such skips indicate that the corresponding sparse sample functions need to be implemented.

In general, there is a little point of defining that an op supports a sparse layout without providing sparse samples (although, there exist a few methods that generate the sparse samples within the methods based on the sparse layout support flag but these methods don't call the `OpInfo.sample_inputs_sparse` method).

The method `_sample_inputs_unspecified` is called only when the following conditions are satisfied all at the same time:
1. An OpInfo instance has specified `supports_sparse_<layout>=True`. Currently, there about 30-40 such OpInfo instances (the exact number depends on the layout).
2. The OpInfo instance has not specified `sample_inputs_sparse_<layout>_func`. Currently, there are 3 OpInfo instances that specify `sample_inputs_sparse_<layout>_func` out of all instances from 1.
3. A `OpInfo.sample_inputs_sparse` method is called. Currently, there are only 4 test methods in `test/test_sparse.py` that call the method.

Atm, the number of OpInfo instances for which all the three conditions hold, is 1 (this is OpInfo for `torch.randn_like`) which leads to `NO SAMPLE FUNCTION SPECIFIED` message when running the `test_like_fns` method from `test_sparse.py`.

**Comment by pearu** [comment]
> Looks good to me, but I'm worried we're adding a ton of skips to our opinfos.

No added skips in op-infos anymore. @cpuhrsch please review.

**Review by cpuhrsch — APPROVED** [review]

**Pull request closed without merging**